src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -83,55 +74,105 @@ private static string GetSchema(ResourceScope targetScope)
             return "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#";
         }
 
-        public void Write(bool emitMetadata, string assemblyVersion, string? templateHash)
+        private readonly JsonTextWriter writer;
+        private readonly EmitterContext context;
+
+        private readonly string assemblyFileVersion;
+
+        public TemplateWriter(JsonTextWriter writer, SemanticModel semanticModel, string assemblyFileVersion)
+        {
+            this.writer = writer;
+            this.context = new EmitterContext(semanticModel);
+            this.assemblyFileVersion = assemblyFileVersion;
+        }
+
+        /// <summary>
+        /// Should be the only constructor used externally
+        /// </summary>
+        public void Write()
+        {
+            this.Write(moduleEmit: false);
+        }
+
+        /// <summary>
+        /// Since emitting modules is a nested call to this, it behaves differently (we reuse the main memoryWriter)
+        /// </summary>
+        public void Write(bool moduleEmit)
```

Comment:
What are your thoughts on:
1. Moving the logic for `moduleEmit == false` directly into the `Write()` overload without params
2. Making this method private, renaming it to `WriteModule()`, and having it just handle the module case (remove the param).

Reply:
Ack! It didn't occur to me that that i could simply rename https://github.com/Azure/bicep/blob/e6919e866727a9302a1e6b252a85cc222a12fddf/src/Bicep.Core/Emit/TemplateWriter.cs#L389
to WriteModule. Nice suggestion!